classes/task/crawl_task.php

```diff
@@ -31,29 +30,55 @@
 
 /**
  * crawl_task
+ * Creates a batch of crawls to be done in an ad hoc task.
+ * This gives us control over how parallel it is, and also when the workload is processed (eg at night).
  *
  * @package    tool_crawler
  * @copyright  2016 Brendan Heywood <brendan@catalyst-au.net>
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 class crawl_task extends \core\task\scheduled_task {
-
     /**
      * Get task name
      */
     public function get_name() {
-        return get_string('pluginname', 'tool_crawler');
+        return get_string('crawl_task', 'tool_crawler');
     }
 
     /**
      * Execute task
      */
     public function execute() {
-        if (\tool_crawler\robot\crawler::get_config()->disablebot === '1') {
+        $config = \tool_crawler\robot\crawler::get_config();
+        if ($config->disablebot === '1') {
             return;
         }
-
-        tool_crawler_crawl();
+        $limitadhoctasks = $config->crawl_task;
+        if (!$limitadhoctasks) {
+            return;
+        }
+        self::tool_crawler_add_adhoc_task($limitadhoctasks);
+    }
+    /**
+     * Add a batch of ad hoc crawl tasks
+     * @param integer $limitadhoctasks the limit of concurrent adhoc crawl tasks
+     * @param boolean $verbose show verbose feedback
+     */
+    public function tool_crawler_add_adhoc_task($limitadhoctasks, $verbose = false) {
+        $crawltask = new \tool_crawler\task\adhoc_crawl_task();
+        $crawltask->set_component('tool_crawler');
+        if ($verbose) {
+            echo "Adding $limitadhoctasks ad hoc tasks to the queue";
+        }
+        // Queue the adhoc_crawl_task $limitadhoctasks times.
+        for ($i = 0; $i < $limitadhoctasks; $i++) {
+            $crawltask->set_custom_data(['worker' => $i]);
+            // We set customdata so that the task API will ignore adding duplicates.
+            \core\task\manager::queue_adhoc_task($crawltask, true); // Need true to check for duplicate workers.
+        }
+        if ($limitadhoctasks < 1 && $verbose) {
+            echo "Already have a full ad hoc task queue with $limitadhoctasks tasks";
```

Comment:
this seems like dead code to me?

Reply:
Yes this doesn't apply now, since we always add all the workers and rely on comparing the customdata to stop duplicates or extras being added